### PR TITLE
[Notifier] Add Sipgate bridge

### PR DESCRIPTION
**Pull request by sakul95**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT

Added a new notifier bridge for sipgate.

### CODE REVIEW

File: src/Symfony/Component/Notifier/Bridge/Sipgate/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+7.1
```
Comment (xabbuh) [inline src/Symfony/Component/Notifier/Bridge/Sipgate/CHANGELOG.md]:
```suggestion
7.2
```

File: src/Symfony/Component/Notifier/Bridge/Sipgate/CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+7.2
+---
+
+ * Added the bridge
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Sipgate/CHANGELOG.md]:
```suggestion
 * Add the bridge
```

File: src/Symfony/Component/Notifier/Bridge/Sipgate/README.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+-----------
+
+```
+SIPGATE_DSN=sipgate://TOKENID:TOKEN@default?senderId=SENDERID
+```
+
+where:
+ - `TOKENID` is your Sipgate API Token ID
+ - `TOKEN` is your Sipgate API TOKEN
+ - `SENDERID` is your Sipgate device id (e.g. s1)
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Sipgate/README.md]:
```suggestion
 - `SENDERID` is your Sipgate device ID (e.g. s1)
```

Comment (OskarStark) [reply src/Symfony/Component/Notifier/Bridge/Sipgate/README.md]: If it is the device ID, why not name it `DEVICE_ID`?

File: src/Symfony/Component/Notifier/Bridge/Sipgate/README.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+DSN example
+-----------
+
+```
+SIPGATE_DSN=sipgate://TOKENID:TOKEN@default?senderId=SENDERID
+```
+
+where:
+ - `TOKENID` is your Sipgate API Token ID
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Sipgate/README.md]:
```suggestion
 - `TOKEN_ID` is your Sipgate API Token ID
```

File: src/Symfony/Component/Notifier/Bridge/Sipgate/README.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
+Sipgate Notifier
+===============
+
+Provides [Sipgate](https://www.sipgate.de) integration for Symfony Notifier.
+
+DSN example
+-----------
+
+```
+SIPGATE_DSN=sipgate://TOKENID:TOKEN@default?senderId=SENDERID
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Sipgate/README.md]:
```suggestion
SIPGATE_DSN=sipgate://TOKEN_ID:TOKEN@default?senderId=SENDERID
```

File: src/Symfony/Component/Notifier/Bridge/Sipgate/SipgateTransport.php [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            throw new TransportException(sprintf('Unable to send SMS with Sipgate: Error code %d - tokenId or token is wrong.', $statusCode), $response);    
+        } elseif (402 === $statusCode) {
+            throw new TransportException(sprintf('Unable to send SMS with Sipgate: Error code %d - insufficient funds.', $statusCode), $response);    
+        } elseif (403 === $statusCode) {
+            throw new TransportException(sprintf('Unable to send SMS with Sipgate: Error code %d - no permisssion to use sms feature or password must be reset or senderId is wrong.', $statusCode), $response);
+        } else {
+            throw new TransportException(sprintf('Unable to send SMS with Sipgate: Error code %d.', $statusCode), $response);    
+        }
+    }
+}
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Sipgate/SipgateTransport.php]: Please add a newline at the end of the file (same for the other files), to make this red pictogram disappear.

You can configure your IDE to do so.

Thanks

File: src/Symfony/Component/Notifier/Bridge/Sipgate/composer.json [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        },
+        {
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": ">=8.2",
+        "symfony/http-client": "^6.4|^7.0",
+        "symfony/notifier": "^6.4|^7.0",
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Sipgate/composer.json]:
```suggestion
        "symfony/notifier": "^7.2",
```

File: src/Symfony/Component/Notifier/Bridge/Sipgate/composer.json [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        "symfony/polyfill-mbstring": "^1.0"
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Sipgate/composer.json]: Not sure this should be in require section, but rather `replace`?

cc @stof

**Review by OskarStark — COMMENTED** [review]
Thank you, I left some comments.

Can you also create a docs and recipes PR?

File: src/Symfony/Component/Notifier/Bridge/Sipgate/README.md (hunk 3 above)
Comment (sakul95) [reply src/Symfony/Component/Notifier/Bridge/Sipgate/README.md]: Actually its called _smsId_ in the official sipgate documentation (https://api.sipgate.com/v2/doc#/sessions/sendWebSms), I just found it confusing and therefore changed it to senderId, as it references to the ID of the sending account.

**Comment by sakul95** [comment]
> Thank you, I left some comments.
> 
> Can you also create a docs and recipes PR?

Yes sure, I have not created a recipe yet so will read into it first.

File: src/Symfony/Component/Notifier/Bridge/Sipgate/SipgateTransport.php (hunk 6 above)
Comment (sakul95) [reply src/Symfony/Component/Notifier/Bridge/Sipgate/SipgateTransport.php]: Done :)

**Comment by sakul95** [comment]
@OskarStark 

Docs: https://github.com/symfony/symfony-docs/pull/20018
Recipe: https://github.com/symfony/recipes/pull/1323

Is this the correct way to do this?

File: src/Symfony/Component/Notifier/Bridge/Sipgate/SipgateTransport.php [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+            return $sentMessage;
+        } elseif (401 === $statusCode) {
+            throw new TransportException(sprintf('Unable to send SMS with Sipgate: Error code %d - tokenId or token is wrong.', $statusCode), $response);    
+        } elseif (402 === $statusCode) {
+            throw new TransportException(sprintf('Unable to send SMS with Sipgate: Error code %d - insufficient funds.', $statusCode), $response);    
+        } elseif (403 === $statusCode) {
+            throw new TransportException(sprintf('Unable to send SMS with Sipgate: Error code %d - no permisssion to use sms feature or password must be reset or senderId is wrong.', $statusCode), $response);
+        } else {
+            throw new TransportException(sprintf('Unable to send SMS with Sipgate: Error code %d.', $statusCode), $response);    
+        }
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Sipgate/SipgateTransport.php]:
```suggestion
        }
        throw new TransportException(sprintf('Unable to send SMS with Sipgate: Error code %d.', $statusCode), $response);
```

File: src/Symfony/Component/Notifier/Bridge/Sipgate/README.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
+Sipgate Notifier
+===============
+
+Provides [Sipgate](https://www.sipgate.de) integration for Symfony Notifier.
+
+DSN example
+-----------
+
+```
+SIPGATE_DSN=sipgate://TOKEN_ID:TOKEN@default?senderId=SENDERID
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Sipgate/README.md]:
```suggestion
SIPGATE_DSN=sipgate://TOKEN_ID:TOKEN@default?senderId=SENDER_ID
```
as this is, what you used in the recipe

File: src/Symfony/Component/Notifier/Bridge/Sipgate/README.md [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+```
+
+where:
+ - `TOKEN_ID` is your Sipgate API Token ID
+ - `TOKEN` is your Sipgate API TOKEN
+ - `SENDERID` is your Sipgate device id (e.g. s1)
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Sipgate/README.md]:
```suggestion
 - `SENDER_ID` is your Sipgate device ID (e.g. s1)
```

**Comment by sakul95** [comment]
@OskarStark Maybe you can answer me two more questions:

1) Is it normal that 3 of the tests fail and if not, what can I do to fix these?

2) The recipe tests fail because the missing package on packagist. Shall I upload the package there or how is the workflow regarding those optional parts of symfony?

Best regards
Lukas

**Comment by fabpot** [comment]
> @OskarStark Maybe you can answer me two more questions:
> 
>     1. Is it normal that 3 of the tests fail and if not, what can I do to fix these?
> 

You can ignore all errors that are not related to your work.

>     2. The recipe tests fail because the missing package on packagist. Shall I upload the package there or how is the workflow regarding those optional parts of symfony?

This will be automatically fixed when I will submit the package. Nothing to do on your side.

> 
> 
> Best regards Lukas

**Comment by xabbuh** [comment]
> Is it normal that 3 of the tests fail and if not, what can I do to fix these?

The failures in the „Verify Packages“ job are related. Looks like some files are missing. You can take inspiration from other existing bridges.

**Comment by fabpot** [comment]
There are some hidden useful information in the broken tests:

```
  File src/Symfony/Component/Notifier/Bridge/Sipgate/.gitattributes does not exist
  File src/Symfony/Component/Notifier/Bridge/Sipgate/.gitignore does not exist
  File src/Symfony/Component/Notifier/Bridge/Sipgate/phpunit.xml.dist does not exist
```

File: src/Symfony/Component/Notifier/Bridge/Sipgate/README.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
+Sipgate Notifier
+===============
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Sipgate/README.md]:
```suggestion
================
```

File: src/Symfony/Component/Notifier/Bridge/Sipgate/composer.json [hunk 13]
```diff
@@ -0,0 +1,32 @@
+{
+    "name": "symfony/sipgate",
```
Comment (xabbuh) [inline src/Symfony/Component/Notifier/Bridge/Sipgate/composer.json]:
```suggestion
    "name": "symfony/sipgate-notifier",
```

**Comment by sakul95** [comment]
I've added the missing files but the _Verify Packages_ check still fails, this time because of "Composer.json's replace section needs to contain symfony/sipgate-notifier". 

I have read into the composer schema and also looked into other comparable packages and can't find a clue why I should do that. 

Of course I could simply add it but that does not seem to be right, maybe you can give me some insight for that?
(Maybe it correlates with comment of @OskarStark  https://github.com/symfony/symfony/pull/57627#discussion_r1663579197)

**Comment by sakul95** [comment]
I've added the missing files but the _Verify Packages_ check still fails, this time because of "Composer.json's replace section needs to contain symfony/sipgate-notifier". 

I have read into the composer schema and also looked into other comparable packages and can't find a clue why I should do that. 

Of course I could simply add it but that does not seem to be right, maybe you can give me some insight for that? @xabbuh

File: src/Symfony/Component/Notifier/Bridge/Sipgate/.gitattributes [hunk 14]
```diff
@@ -0,0 +1,4 @@
+/Tests export-ignore
+/phpunit.xml.dist export-ignore
+/.gitattributes export-ignore
+/.gitignore export-ignore
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Sipgate/.gitattributes]:
```suggestion
/.git* export-ignore
```

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @sakul95.

**Pull request merged**